test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -0,0 +1,372 @@
+ * generate_samples should throw no errors. For all combinations with an invalid
+ * (bad) parameter, generate_samples should throw domain_errors.
+ *
+ * The generate_samples callable is also passed various other guaranteed invalid
+ * values like positive infinity, negative infinity, and NaNs (these should also
+ * cause domain_errors).
+ *
+ * The generated_samples callable is also tested to reject incompatibly sized
+ * input vectors. These should cause invalid_argument errors.
+ *
+ * The generate_samples callable must have the signature:
+ *   T_out generate_samples(T_param1 p1, T_param2 p2, T_rng)
+ * where T_param1 and T_param2 can be any combination of:
+ *   double, std::vector<double>, VectorXd, or RowVectorXd
+ * and T_rng is a boost random number generator
+ *
+ * The output of generate_samples must be of
+ *   stan::length(out) == stan::max_size(p1, p2)
+ *
+ * Each element of the output, out[i], should correspond to a random variate
+ * generated according to parameters p1[i] and p2[i]
+ *
+ * @tparam T_param1 Type of first parameter
+ * @tparam T_param2 Type of second parameter
+ * @tparam T_generate_samples Type of generate_samples functor
+ * @param generate_samples generate_samples functor to test
+ * @param good_p1_ Valid values of first parameter
+ * @param bad_p1_ Invalid values of first parameter
+ * @param good_p2_ Valid values of second parameter
+ * @param bad_p2_ Invalid values of second parameter
+ */
+template<typename T_param1, typename T_param2, typename T_generate_samples>
+void check_dist_throws(T_generate_samples generate_samples,
+                       const std::vector<double>& good_p1_,
+                       const std::vector<double>& bad_p1_,
+                       const std::vector<double>& good_p2_,
+                       const std::vector<double>& bad_p2_) {
+  boost::random::mt19937 rng;
+
+  T_param1 p1;
+  T_param2 p2;
+
+  resize_if_vector(p1, 5);  // No-op if p1 is a scalar
+  resize_if_vector(p2, 5);  // No-op if p2 is a scalar
+
+  // Make copies of the input arguments so that we can randomly shuffle them
+  // in the tests
+  std::vector<double> good_p1 = good_p1_;
+  std::vector<double> bad_p1 = bad_p1_;
+  std::vector<double> good_p2 = good_p2_;
+  std::vector<double> bad_p2 = bad_p2_;
+
+  // Try a few combinations of parameters that should work
+  for (int i = 0; i < 5; i++) {
+    std::random_shuffle(good_p1.begin(), good_p1.end());
+    std::random_shuffle(good_p2.begin(), good_p2.end());
+    assign_parameter_values(p1, good_p1);
+    assign_parameter_values(p2, good_p2);
+    EXPECT_NO_THROW(generate_samples(p1, p2, rng));
+  }
+
+  // Now try putting incompatible values in first parameter
+  if (bad_p1.size() > 0) {
+    for (int i = 0; i < 5; i++) {
+      std::random_shuffle(bad_p1.begin(), bad_p1.end());
+      std::random_shuffle(good_p2.begin(), good_p2.end());
+      assign_parameter_values(p1, bad_p1);
```

Comment:
I don't think it matters that I'm filling the thing completely with bad values, though I could add a set_one_value function. I just reused assign_parameter_values to avoid too many weird helper functions.

But shuffling the bad values is just silly. I need to ensure specifically that each one causes a fail. I should just loop through the bad values. It's quite possible I could miss one this way.

Reply:
Yeah, it would be good if our tests to be deterministic when possible :P